### PR TITLE
drivers: serial: added UART interrupts for Raspberry Pi Pico board

### PR DESCRIPTION
**Pull request by andrei-edward-popa**
Added UART Interrupt Driven driver for Raspberry Pi Pico board.

Signed-off-by: Andrei-Edward Popa <andrei_edward.popa@upb.ro>

### CODE REVIEW

File: drivers/serial/uart_rpi_pico.c [hunk 1]
```diff
@@ -14,62 +15,313 @@
...
 };
 
+#define DEV_CFG(dev)  \
+	((const struct uart_rpi_config *const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct uart_rpi_data *const)(dev)->data)
+#define DEV_UART_INST(dev) \
+	((uart_inst_t *const)((const struct uart_rpi_config *const)(dev)->config)->uart_dev)
+#define DEV_UART_HW(dev) \
+	uart_get_hw(DEV_UART_INST(dev))
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:39]: please remove this sort of accessors, ref. https://github.com/zephyrproject-rtos/zephyr/pull/41918

File: drivers/serial/uart_rpi_pico.c [hunk 2]
```diff
@@ -14,62 +15,313 @@
...
+	};	\
+	\
+	RPI_UART_CONFIG_FUNC(idx)	\
+	RPI_UART_CREATE_CFG(idx);	\
+	\
+	DEVICE_DT_INST_DEFINE(idx, &uart_rpi_init,	\
+			    NULL, &uart##idx##_rpi_data,	\
+			    &uart##idx##_rpi_config, PRE_KERNEL_1,	\
+			    CONFIG_SERIAL_INIT_PRIORITY,	\
+			    &uart_rpi_driver_api);	\
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:325]: please align `\` to the same column for readability

File: drivers/serial/uart_rpi_pico.c [hunk 3]
```diff
@@ -14,62 +15,313 @@
...
+			    DEVICE_DT_INST_GET(idx), 0);	\
+		irq_enable(DT_INST_IRQN(idx));	\
+	}
+
+#else
+
+#define RPI_UART_CREATE_CFG(idx)	\
+	static const struct uart_rpi_config uart##idx##_rpi_config = {	\
+		.uart_dev = uart##idx,	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:305]: this is duplicating initializer fields unnecessarily, better use something like:

```
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
#define RPI_UART_IRQ_CONFIG_INIT .irq_config_func = ...,
#else
#define RPI_UART_IRQ_CONFIG_INIT
#endif
```

or, make use of `IF_ENABLED(...)`

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_rpi_pico.c (hunk 1 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Ok, I removed all accessors.

File: drivers/serial/uart_rpi_pico.c (hunk 3 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Done!

File: drivers/serial/uart_rpi_pico.c (hunk 2 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: I aligned it before, but the pipeline check said that I have a line with more than 100 characters, but it was only 77 characters long. I will try to align it in a different way.

Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: I found that the tab width was the problem. Now should be okay, you can check it.

File: drivers/serial/uart_rpi_pico.c [hunk 4]
```diff
@@ -14,62 +15,307 @@
...
 
 static int uart_rpi_poll_in(const struct device *dev, unsigned char *c)
 {
-	const struct uart_rpi_config *config = dev->config;
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	uart_hw_t * const uart_hw = uart_get_hw(uart_inst);
 
-	if (!uart_is_readable(config->uart_dev)) {
+	if (!uart_is_readable(uart_inst)) {
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:38]: why is this change needed? it doesn't seem related to interrupt support.

File: boards/arm/rpi_pico/rpi_pico.dts [hunk 5]
```diff
@@ -14,6 +14,7 @@
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
 		zephyr,console = &uart0;
+		zephyr,shell-uart = &uart0;
```
Comment (gmarull) [inline boards/arm/rpi_pico/rpi_pico.dts:17]: please add board changes to another commit

File: drivers/serial/uart_rpi_pico.c (hunk 4 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:38]: This change is not needed, sorry for that.

File: boards/arm/rpi_pico/rpi_pico.dts (hunk 5 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico.dts:17]: Okay.

File: drivers/serial/uart_rpi_pico.c [hunk 6]
```diff
@@ -14,62 +15,307 @@
...
+{
+	int tx_len = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	uart_hw_t * const uart_hw = uart_get_hw(uart_inst);
+
+	while (uart_is_writable(uart_inst) && (len - tx_len) > 0) {
+		uart_hw->dr = tx_data[tx_len++];
+	}
```
[the remark this answers is not in the document]
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:130]: Yes, uart_hw_t is the register structure that can be accessed by uart_get_hw API function. I will call this function anytime I need, I will remove uart_hw pointer from everywhere.

File: drivers/serial/uart_rpi_pico.c [hunk 7]
```diff
@@ -14,62 +15,307 @@
...
 	if (ret < 0) {
 		return ret;
 	}
 
-	uart_init(config->uart_dev, config->baudrate);
+	baudrate = uart_init(uart_inst, data->baudrate);
+	/* Check if baudrate adjustment returned by 'uart_init' function is a positive value */
+	if (baudrate <= 0) {
+		return -EINVAL;
+	}
```
[the remark this answers is not in the document]
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:68]: I added a separate commit for this modification.

File: drivers/serial/uart_rpi_pico.c [hunk 8]
```diff
@@ -70,7 +74,7 @@ static int uart_rpi_init(const struct device *dev)
 	config->irq_config_func(dev);
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
-	return 0;
+	return ret;
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:77]: this can stay to `0`

File: drivers/serial/Kconfig.rpi_pico [hunk 9]
```diff
@@ -9,3 +9,4 @@ config UART_RPI_PICO
 	default $(dt_compat_enabled,$(DT_COMPAT_RPI_PICO_UART))
 	select SERIAL_HAS_DRIVER
 	select PICOSDK_USE_UART
+	select SERIAL_SUPPORT_INTERRUPT
```
Comment (gmarull) [inline drivers/serial/Kconfig.rpi_pico:12]: this is up to the board to decide, I guess

File: boards/arm/rpi_pico/rpi_pico_defconfig [hunk 10]
```diff
@@ -7,3 +7,4 @@ CONFIG_UART_CONSOLE=y
 CONFIG_GPIO=y
 CONFIG_USE_DT_CODE_PARTITION=y
 CONFIG_BUILD_OUTPUT_UF2=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
```
Comment (gmarull) [inline boards/arm/rpi_pico/rpi_pico_defconfig:10]: this is updating board: should go into a separate commit

File: drivers/serial/uart_rpi_pico.c [hunk 11]
```diff
@@ -39,37 +50,249 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
...
+
+static void uart_rpi_irq_err_enable(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+
+	uart_get_hw(config->uart_dev)->imsc |= (UART_UARTIMSC_OEIM_BITS |
+											UART_UARTIMSC_BEIM_BITS |
+											UART_UARTIMSC_PEIM_BITS |
+											UART_UARTIMSC_FEIM_BITS |
+											UART_UARTIMSC_RTIM_BITS);
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:200]: looks like formatting needs to be improved here

File: drivers/serial/uart_rpi_pico.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/serial/uart_rpi_pico.c]: my initial concern has not been addressed: with these changes, the driver is mixing high-level APIs with register access, which, I think, is confusing. If high-level API is not suitable for Zephyr drivers, then everything should be moved to register based for consistency.

File: boards/arm/rpi_pico/rpi_pico_defconfig (hunk 10 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico_defconfig:10]: Can be in the same commit with dts modifications?

File: drivers/serial/uart_rpi_pico.c (hunk 8 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Okay, I will let it 0.

File: drivers/serial/Kconfig.rpi_pico (hunk 9 above)
Comment (andrei-edward-popa) [reply drivers/serial/Kconfig.rpi_pico:12]: In all drivers that I saw in zephyr this was here.

File: boards/arm/rpi_pico/rpi_pico_defconfig (hunk 10 above)
Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico_defconfig:10]: Done!

File: drivers/serial/uart_rpi_pico.c (hunk 11 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Yes, I formatted it.

File: drivers/serial/uart_rpi_pico.c [hunk 12]
```diff
@@ -53,27 +66,238 @@ static int uart_rpi_init(const struct device *dev)
...
+
+	return errors;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static int uart_rpi_fifo_fill(const struct device *dev,
+			      const uint8_t *tx_data, int len)
+{
+	int tx_len = 0;
```
Comment (pirearadu) [inline drivers/serial/uart_rpi_pico.c:111]: Please use reverse christmas tree declaration.

File: drivers/serial/uart_rpi_pico.c [hunk 13]
```diff
@@ -53,27 +66,238 @@ static int uart_rpi_init(const struct device *dev)
...
+
+	return tx_len;
+}
+
+static int uart_rpi_fifo_read(const struct device *dev,
+			      uint8_t *rx_data, const int len)
+{
+	int rx_len = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
```
Comment (pirearadu) [inline drivers/serial/uart_rpi_pico.c:128]: Please use reverse christmas tree declaration.

File: drivers/serial/uart_rpi_pico.c (hunk 12 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:111]: Okay, I will modify everywhere.

File: drivers/serial/uart_rpi_pico.c [hunk 14]
```diff
@@ -53,27 +66,238 @@ static int uart_rpi_init(const struct device *dev)
...
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
 	return 0;
 }
 
+static int uart_rpi_err_check(const struct device *dev)
+{
+	int errors = 0;
+
+	const struct uart_rpi_config * const config = dev->config;
```
Comment (pirearadu) [inline drivers/serial/uart_rpi_pico.c:83]: Please use reverse christmas tree declaration.
And no space between declarations.

File: drivers/serial/uart_rpi_pico.c (hunk 13 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:123]: Done!

File: drivers/serial/uart_rpi_pico.c (hunk 14 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:81]: Done!

File: drivers/serial/Kconfig.rpi_pico (hunk 9 above)
Comment (gmarull) [reply drivers/serial/Kconfig.rpi_pico:12]: sorry, got confused with `CONFIG_UART_INTERRUPT_DRIVEN`, my fault.

File: boards/arm/rpi_pico/rpi_pico_defconfig (hunk 10 above)
Comment (gmarull) [reply boards/arm/rpi_pico/rpi_pico_defconfig:10]: nit: for bisectability this commit needs to go after adding interrupt support

Comment (andrei-edward-popa) [reply boards/arm/rpi_pico/rpi_pico_defconfig:10]: Okay, makes sense.

File: drivers/serial/uart_rpi_pico.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: I followed the approach from the polling method right now.

File: drivers/serial/uart_rpi_pico.c [hunk 15]
```diff
@@ -40,36 +51,249 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
...
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static int uart_rpi_fifo_fill(const struct device *dev,
+			      const uint8_t *tx_data, int len)
+{
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	int tx_len = 0;
+
+	while (uart_is_writable(uart_inst) && (len - tx_len) > 0) {
```
Comment (gmarull) [inline drivers/serial/uart_rpi_pico.c:113]: this driver continues to mix API and register access. I think it would be better to just use register access everywhere, so for example:

```suggestion
	while (!(uart_get_hw(uart)->fr & UART_UARTFR_TXFF_BITS) && (len - tx_len) > 0) {
```

`uart_get_hw` could actually be removed by just storing a `uart_hw_t *`.

Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:113]: Okay, I understand now, I will do this modifications soon.

Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:113]: @gmarull What about uart_init function?

Comment (gmarull) [reply drivers/serial/uart_rpi_pico.c]: note: `uart_init()` actions should also be handled by the driver

Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Note that uart_init function is using reset and clock API functions, but reset driver and clock driver are not implemented in Zephyr yet. If I want to implement the init function I need to bring another API functions. I think for now this API call is not dangerous and in the future I would like to implement those drivers in Zephyr and then I will modify also this function.

Comment (gmarull) [reply drivers/serial/uart_rpi_pico.c]: Fair enough, I guess it's ok for now then.

Comment (gmarull) [reply drivers/serial/uart_rpi_pico.c]: Would you mind adding a comment around `uart_init()`?

**Review by gmarull — APPROVED** [review]

File: drivers/serial/uart_rpi_pico.c (hunk 15 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: A comment like?

Comment (gmarull) [reply drivers/serial/uart_rpi_pico.c]:
```
/* uart_init() may be replaced by register based API once rpi-pico platform has a clock controller driver */
```

Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Done! Thanks for review. :) Tomorrow I will make a PR with I2C driver.

**Comment by andrei-edward-popa** [comment]
@dcpleung Hi! Can you review my code? :)

File: drivers/serial/uart_rpi_pico.c [hunk 16]
```diff
@@ -40,6 +52,9 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
 static int uart_rpi_init(const struct device *dev)
 {
 	const struct uart_rpi_config *config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
```
Comment (dcpleung) [inline drivers/serial/uart_rpi_pico.c:61]: To keep the tree bisect-able, this should be in the previous commit where the call to `uart_init()` was added.

File: drivers/serial/uart_rpi_pico.c [hunk 17]
```diff
@@ -40,6 +52,9 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
...
+	uart_hw_t * const uart_hw = config->uart_regs;
+	struct uart_rpi_data * const data = dev->data;
```
Comment (dcpleung) [inline drivers/serial/uart_rpi_pico.c:63]: This also needs to be in the previous commit.

**Review by dcpleung — CHANGES_REQUESTED** [review]
LGTM but please move the introduction of variables to the correct commit.

File: drivers/serial/uart_rpi_pico.c (hunk 17 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:63]: Done!

File: drivers/serial/uart_rpi_pico.c (hunk 16 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:61]: Yes, you are right, it was my mistake. It is okay now.

**Review by dcpleung — APPROVED** [review]
